{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "from googleapiclient.discovery import build\n",
    "import json\n",
    "from dateutil import relativedelta\n",
    "from github import Github\n",
    "import requests\n",
    "from urllib.parse import quote\n",
    "import os\n",
    "import zipfile\n",
    "from io import BytesIO\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Important Links and APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "training_url = 'https://training.galaxyproject.org/training-material/tags/microgalaxy/embed.html'\n",
    "training_material_api = ''\n",
    "youtube_api_key = '' #'YOUR_YOUTUBE_API_KEY'\n",
    "plausible_api_key = \"\"  #'YOUR_PLAUSIBLE_API_KEY'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## microGalaxy GTN Youtube Videos Stats Extraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions extracting microGalaxy training material names, information and plausible links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to fetch and parse the HTML content\n",
    "def fetch_and_parse_html(url):\n",
    "    response = requests.get(url)\n",
    "    soup = BeautifulSoup(response.content, 'html.parser')\n",
    "    return soup\n",
    "\n",
    "# Function to extract table data and create a DataFrame\n",
    "def extract_table_data(soup):\n",
    "    tables = soup.find_all('table')\n",
    "    \n",
    "    all_rows = []\n",
    "    for table in tables:\n",
    "        headers = [header.text.strip() for header in table.find_all('th')]\n",
    "        rows = []\n",
    "        for row in table.find_all('tr')[1:]:\n",
    "            row_data = [data.text.strip() for data in row.find_all('td')]\n",
    "            rows.append(row_data)\n",
    "        all_rows.extend(rows)\n",
    "    \n",
    "    df = pd.DataFrame(all_rows, columns=headers)\n",
    "    return df\n",
    "\n",
    "# Function to remove 'Hands-on: Hands-on: ' from Lesson column entries\n",
    "def clean_lesson_column(df):\n",
    "    df['Lesson'] = df['Lesson'].str.replace('Hands-on: Hands-on: ', '', regex=False)\n",
    "    df['Lesson'] = df['Lesson'].str.split('\\n').str[0]\n",
    "    return df\n",
    "\n",
    "# Function to create a boolean column indicating if 'Recordings' is present\n",
    "def create_recordings_present_column(df):\n",
    "    df['Recordings_Present'] = df['Recordings'].str.startswith('video\\n')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions to extract youtube stats, checking the GalaxyProject channel, for every microGalaxy training and return the video date and number of views"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: QIIME 2 Cancer Microbiome Intervention external-link GalaxyProject site:youtube.com\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Lesson</th>\n",
       "      <th>Recordings_Present</th>\n",
       "      <th>Video_Count</th>\n",
       "      <th>Published_Date</th>\n",
       "      <th>Video_ID</th>\n",
       "      <th>Video_URL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Assembly of metagenomic sequencing data</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Unicycler Assembly</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Identifying tuberculosis transmission links: f...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tree thinking for tuberculosis evolution and e...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bacterial Genome Annotation</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Genome annotation with Prokka</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Refining Genome Annotations with Apollo (proka...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Comparative gene analysis in unannotated genomes</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>16S Microbial Analysis with mothur (extended)</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>16S Microbial Analysis with mothur (short)</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>16S Microbial analysis with Nanopore data</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Analyses of metagenomics data - The global pic...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Antibiotic resistance detection</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Binning of metagenomic sequencing data</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Identification of the micro-organisms in a bee...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Metatranscriptomics analysis using microbiome ...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Metatranscriptomics analysis using microbiome ...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Pathogen detection from (direct Nanopore) sequ...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>QIIME 2 Cancer Microbiome Intervention externa...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Taxonomic Profiling and Visualization of Metag...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Metaproteomics tutorial</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>metaQuantome 1: Data creation</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>metaQuantome 2: Function</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>metaQuantome 3: Taxonomy</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Calling variants in non-diploid systems</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>M. tuberculosis Variant Analysis</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Microbial Variant Calling</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Lesson  Recordings_Present  \\\n",
       "0             Assembly of metagenomic sequencing data               False   \n",
       "1   Genome Assembly of a bacterial genome (MRSA) s...               False   \n",
       "2            Making sense of a newly assembled genome               False   \n",
       "3                                  Unicycler Assembly               False   \n",
       "4   Identifying tuberculosis transmission links: f...               False   \n",
       "5   Tree thinking for tuberculosis evolution and e...               False   \n",
       "6                         Bacterial Genome Annotation               False   \n",
       "7                       Genome annotation with Prokka                True   \n",
       "8   Refining Genome Annotations with Apollo (proka...                True   \n",
       "9    Comparative gene analysis in unannotated genomes               False   \n",
       "10      16S Microbial Analysis with mothur (extended)               False   \n",
       "11         16S Microbial Analysis with mothur (short)               False   \n",
       "12          16S Microbial analysis with Nanopore data               False   \n",
       "13  Analyses of metagenomics data - The global pic...               False   \n",
       "14                    Antibiotic resistance detection               False   \n",
       "16             Binning of metagenomic sequencing data               False   \n",
       "17  Identification of the micro-organisms in a bee...               False   \n",
       "18  Metatranscriptomics analysis using microbiome ...               False   \n",
       "19  Metatranscriptomics analysis using microbiome ...               False   \n",
       "20  Pathogen detection from (direct Nanopore) sequ...               False   \n",
       "21  QIIME 2 Cancer Microbiome Intervention externa...               False   \n",
       "22  Taxonomic Profiling and Visualization of Metag...               False   \n",
       "23                            Metaproteomics tutorial                True   \n",
       "24                      metaQuantome 1: Data creation               False   \n",
       "25                           metaQuantome 2: Function               False   \n",
       "26                           metaQuantome 3: Taxonomy               False   \n",
       "27            Calling variants in non-diploid systems               False   \n",
       "28                   M. tuberculosis Variant Analysis                True   \n",
       "29                          Microbial Variant Calling               False   \n",
       "\n",
       "   Video_Count Published_Date Video_ID Video_URL  \n",
       "0         None                    None      None  \n",
       "1         None                    None      None  \n",
       "2         None                    None      None  \n",
       "3         None                    None      None  \n",
       "4         None                    None      None  \n",
       "5         None                    None      None  \n",
       "6         None                    None      None  \n",
       "7         None                    None      None  \n",
       "8         None                    None      None  \n",
       "9         None                    None      None  \n",
       "10        None                    None      None  \n",
       "11        None                    None      None  \n",
       "12        None                    None      None  \n",
       "13        None                    None      None  \n",
       "14        None                    None      None  \n",
       "16        None                    None      None  \n",
       "17        None                    None      None  \n",
       "18        None                    None      None  \n",
       "19        None                    None      None  \n",
       "20        None                    None      None  \n",
       "21        None           None     None      None  \n",
       "22        None                    None      None  \n",
       "23        None                    None      None  \n",
       "24        None                    None      None  \n",
       "25        None                    None      None  \n",
       "26        None                    None      None  \n",
       "27        None                    None      None  \n",
       "28        None                    None      None  \n",
       "29        None                    None      None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Function to get video details for each lesson with channel, title, and order check\n",
    "def get_video_details(api_key, video_titles, channel_id='UCx3M4_d3M__lXf9OX8I5Kdg', cache_file='video_cache.json'):\n",
    "    try:\n",
    "        with open(cache_file, 'r') as file:\n",
    "            cache = json.load(file)\n",
    "    except FileNotFoundError:\n",
    "        cache = {}\n",
    "\n",
    "    youtube = build('youtube', 'v3', developerKey=api_key)\n",
    "    video_details = {}\n",
    "\n",
    "    for title in video_titles:\n",
    "        if title in cache:\n",
    "            video_details[title] = cache[title]\n",
    "        else:\n",
    "            # Print the query\n",
    "            query = f'{title} GalaxyProject site:youtube.com'\n",
    "            print(f'Query: {query}')\n",
    "\n",
    "            # Execute the YouTube API request\n",
    "            request = youtube.search().list(\n",
    "                part='id,snippet',\n",
    "                q=query,\n",
    "                type='video',\n",
    "                maxResults=1\n",
    "            )\n",
    "            response = request.execute()\n",
    "\n",
    "            items = response.get('items', [])\n",
    "            if items:\n",
    "                snippet = items[0]['snippet']\n",
    "                video_id = items[0]['id']['videoId']\n",
    "                video_url = f'https://www.youtube.com/watch?v={video_id}'\n",
    "               \n",
    "\n",
    "                # Check if the video is from the specified channel and has a title exactly matching the Lesson entry\n",
    "                if snippet['channelId'] == channel_id and title.lower() in snippet['title'].lower():\n",
    "                    video_details[title] = {\n",
    "                        'Video_Count': int(snippet.get('statistics', {}).get('viewCount', 0)),\n",
    "                        'Published_Date': snippet.get('publishedAt', ''),\n",
    "                        'Video_ID': video_id,\n",
    "                        'Video_URL': video_url\n",
    "                    }\n",
    "                else:\n",
    "                    # Video doesn't meet the criteria, set Video_Count to None\n",
    "                    video_details[title] = {\n",
    "                        'Video_Count': None,\n",
    "                        'Published_Date': '',\n",
    "                        'Video_ID': None,\n",
    "                        'Video_URL': None\n",
    "                    }\n",
    "\n",
    "                # Update cache\n",
    "                cache[title] = video_details[title]\n",
    "\n",
    "    # Save updated cache to file\n",
    "    with open(cache_file, 'w') as file:\n",
    "        json.dump(cache, file)\n",
    "\n",
    "    return video_details\n",
    "\n",
    "# Adding youtube videos statistics\n",
    "def creating_trainings_videos_stats():\n",
    "\n",
    "    # Fetch and parse HTML\n",
    "    soup = fetch_and_parse_html(training_url)\n",
    "    \n",
    "    # Extract table data and create a DataFrame\n",
    "    df = extract_table_data(soup)\n",
    "    \n",
    "    # Clean the 'Lesson' column\n",
    "    df_cleaned_lesson = clean_lesson_column(df)\n",
    "    \n",
    "    # Create a boolean column indicating if 'Recordings' is present\n",
    "    df_final = create_recordings_present_column(df_cleaned_lesson)\n",
    "    \n",
    "    # Get video details for each lesson with channel, title, and order check\n",
    "    video_titles = df_final['Lesson'].tolist()\n",
    "    video_details = get_video_details(youtube_api_key, video_titles)\n",
    "    \n",
    "    # Create new columns 'Video_Count', 'Published_Date', 'Video_ID', and 'Video_URL'\n",
    "    df_final['Video_Count'] = df_final['Lesson'].apply(lambda title: video_details.get(title, {}).get('Video_Count'))\n",
    "    df_final['Published_Date'] = df_final['Lesson'].apply(lambda title: video_details.get(title, {}).get('Published_Date'))\n",
    "    df_final['Video_ID'] = df_final['Lesson'].apply(lambda title: video_details.get(title, {}).get('Video_ID'))\n",
    "    df_final['Video_URL'] = df_final['Lesson'].apply(lambda title: video_details.get(title, {}).get('Video_URL'))\n",
    "    \n",
    "    # Drop duplicates based on the 'Lesson' column\n",
    "    df_final = df_final.drop_duplicates(subset=['Lesson'])\n",
    "\n",
    "    # Keep only the desired columns\n",
    "    df_final = df_final[['Lesson', 'Recordings_Present', 'Video_Count', 'Published_Date', 'Video_ID', 'Video_URL']]\n",
    "   \n",
    "    # Display the final DataFrame\n",
    "    display(df_final)\n",
    "\n",
    "# Creating training videos statisitics table\n",
    "creating_trainings_videos_stats()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## microGalaxy GTN visitors and feedback (Plausible)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For /training-material/topics/assembly/tutorials/metagenomics-assembly/tutorial.html, Number of Visitors: 1795696\n",
      "For /training-material/topics/assembly/tutorials/mrsa-illumina/tutorial.html, Number of Visitors: 1795696\n",
      "For /training-material/topics/assembly/tutorials/ecoli_comparison/tutorial.html, Number of Visitors: 1795696\n",
      "For /training-material/topics/assembly/tutorials/unicycler-assembly/tutorial.html, Number of Visitors: 1795696\n",
      "For /training-material/topics/evolution/tutorials/mtb_transmission/tutorial.html, Number of Visitors: 1795696\n",
      "For /training-material/topics/evolution/tutorials/mtb_phylogeny/tutorial.html, Number of Visitors: 1795697\n",
      "For /training-material/topics/genome-annotation/tutorials/bacterial-genome-annotation/tutorial.html, Number of Visitors: 1795697\n",
      "For /training-material/topics/genome-annotation/tutorials/annotation-with-prokka/tutorial.html, Number of Visitors: 1795697\n",
      "For /training-material/topics/genome-annotation/tutorials/apollo/tutorial.html, Number of Visitors: 1795697\n",
      "For /training-material/topics/genome-annotation/tutorials/gene-centric/tutorial.html, Number of Visitors: 1795697\n",
      "For /training-material/topics/microbiome/tutorials/mothur-miseq-sop/tutorial.html, Number of Visitors: 1795697\n",
      "For /training-material/topics/microbiome/tutorials/mothur-miseq-sop-short/tutorial.html, Number of Visitors: 1795698\n",
      "For /training-material/topics/microbiome/tutorials/nanopore-16S-metagenomics/tutorial.html, Number of Visitors: 1795698\n",
      "For /training-material/topics/microbiome/tutorials/general-tutorial/tutorial.html, Number of Visitors: 1795698\n",
      "For /training-material/topics/microbiome/tutorials/plasmid-metagenomics-nanopore/tutorial.html, Number of Visitors: 1795698\n",
      "For /training-material/topics/microbiome/tutorials/metagenomics-assembly/tutorial.html, Number of Visitors: 1795698\n",
      "For /training-material/topics/microbiome/tutorials/metagenomics-binning/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/microbiome/tutorials/beer-data-analysis/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/microbiome/tutorials/metatranscriptomics/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/microbiome/tutorials/metatranscriptomics-short/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/microbiome/tutorials/pathogen-detection-from-nanopore-foodborne-data/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/microbiome/tutorials/taxonomic-profiling/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/proteomics/tutorials/metaproteomics/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/proteomics/tutorials/metaquantome-data-creation/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/proteomics/tutorials/metaquantome-function/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/proteomics/tutorials/metaquantome-taxonomy/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/variant-analysis/tutorials/non-dip/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.html, Number of Visitors: 1795699\n",
      "For /training-material/topics/variant-analysis/tutorials/microbial-variants/tutorial.html, Number of Visitors: 1795699\n"
     ]
    }
   ],
   "source": [
    "# Get the current date\n",
    "current_date = datetime.now().strftime('%Y-%m-%d')\n",
    "\n",
    "# Send a GET request to the URL\n",
    "response = requests.get(training_url)\n",
    "\n",
    "# Parse the HTML content using BeautifulSoup\n",
    "soup = BeautifulSoup(response.text, 'html.parser')\n",
    "\n",
    "# Extract lesson links\n",
    "lesson_links = [a['href'] for a in soup.find_all('a', href=True)]\n",
    "\n",
    "# Base URL for plausible\n",
    "plausible_base_url = \"https://plausible.galaxyproject.eu\"\n",
    "\n",
    "# Filter only the lesson links that end with \"tutorial.html\"\n",
    "lesson_links = [link for link in lesson_links if link.endswith(\"tutorial.html\")]\n",
    "\n",
    "# Dictionary to store plausible download links for each lesson\n",
    "plausible_download_links = {}\n",
    "\n",
    "# Rest of the code remains unchanged...\n",
    "\n",
    "# Generate plausible download links for the filtered lesson links using the current date\n",
    "for link in lesson_links:\n",
    "    plausible_download_link = f\"{plausible_base_url}/training.galaxyproject.org/export?period=all&date={current_date}&filters=%7B%22page%22%3A%22{quote(link)}%22%7D&with_imported=true&interval=month\"\n",
    "    plausible_download_links[link] = plausible_download_link\n",
    "\n",
    "\n",
    "# Directory to store downloaded zip files\n",
    "download_directory = \"plausible_zips\"\n",
    "\n",
    "# Create the directory if it doesn't exist\n",
    "os.makedirs(download_directory, exist_ok=True)\n",
    "\n",
    "# Iterate through each lesson and download the plausible zip file\n",
    "for lesson, download_link in plausible_download_links.items():\n",
    "    # Properly encode the URL\n",
    "    encoded_link = quote(download_link, safe=':/?&=')\n",
    "\n",
    "    # Specify cache control headers to avoid caching\n",
    "    headers = {\n",
    "        \"Authorization\": f\"Bearer {plausible_api_key}\",\n",
    "        \"Cache-Control\": \"no-cache, no-store, must-revalidate\",\n",
    "        \"Pragma\": \"no-cache\",\n",
    "        \"Expires\": \"0\"\n",
    "    }\n",
    "\n",
    "    # Make the request with the encoded URL and cache control headers\n",
    "    response = requests.get(encoded_link, headers=headers)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        # Generate a unique name for the zip file\n",
    "        zip_filename = f\"{lesson.replace('/', '_')}_{current_date}_plausible.zip\"\n",
    "        zip_filepath = os.path.join(download_directory, zip_filename)\n",
    "        \n",
    "        # Save the zip file with the unique filename\n",
    "        with open(zip_filepath, 'wb') as zip_file:\n",
    "            zip_file.write(response.content)\n",
    "        \n",
    "        # Extract visitors.csv from the zip file\n",
    "        with zipfile.ZipFile(zip_filepath, 'r') as zip_ref:\n",
    "            # Get the name of the visitors.csv file in the zip\n",
    "            csv_name = [name for name in zip_ref.namelist() if name.lower().endswith('visitors.csv')][0]\n",
    "            \n",
    "            # Generate a new name for the visitors.csv file\n",
    "            new_csv_name = f\"{lesson.replace('/', '_')}_{current_date}_visitors.csv\"\n",
    "            \n",
    "            # Extract the visitors.csv file with the new name\n",
    "            zip_ref.extract(csv_name, download_directory)\n",
    "            \n",
    "            # Rename the visitors.csv file\n",
    "            os.rename(os.path.join(download_directory, csv_name), os.path.join(download_directory, new_csv_name))\n",
    "            \n",
    "            # Read visitors.csv into a DataFrame\n",
    "            visitors_df = pd.read_csv(os.path.join(download_directory, new_csv_name))\n",
    "        \n",
    "        # Print the number of visitors for the lesson\n",
    "        print(f\"For {lesson}, Number of Visitors: {visitors_df['visitors'].sum()}\")\n",
    "    else:\n",
    "        print(f\"Error fetching Plausible metrics for {lesson}: {response.status_code}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating Plausible links for every training on GTN with microGalaxy tag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/genome-annotation/tutorials/apollo/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/mothur-miseq-sop/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/general-tutorial/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/genome-annotation/tutorials/bacterial-genome-annotation/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/assembly/tutorials/mrsa-illumina/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/proteomics/tutorials/metaquantome-taxonomy/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/pathogen-detection-from-nanopore-foodborne-data/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/evolution/tutorials/mtb_transmission/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/assembly/tutorials/ecoli_comparison/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/variant-analysis/tutorials/non-dip/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/genome-annotation/tutorials/annotation-with-prokka/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/nanopore-16S-metagenomics/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/assembly/tutorials/metagenomics-assembly/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/proteomics/tutorials/metaquantome-function/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/evolution/tutorials/mtb_phylogeny/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/taxonomic-profiling/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/plasmid-metagenomics-nanopore/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/beer-data-analysis/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/proteomics/tutorials/metaproteomics/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/assembly/tutorials/unicycler-assembly/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/proteomics/tutorials/metaquantome-data-creation/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/metatranscriptomics/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/genome-annotation/tutorials/gene-centric/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/mothur-miseq-sop-short/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/metatranscriptomics-short/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/metagenomics-binning/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/variant-analysis/tutorials/microbial-variants/tutorial.html\n",
      "https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=/training-material/topics/microbiome/tutorials/metagenomics-assembly/tutorial.html\n"
     ]
    }
   ],
   "source": [
    "# URL of the training page\n",
    "training_url = \"https://training.galaxyproject.org/training-material/tags/microgalaxy/embed.html\"\n",
    "\n",
    "# Send a GET request to the URL\n",
    "response = requests.get(training_url)\n",
    "\n",
    "# Parse the HTML content using BeautifulSoup\n",
    "soup = BeautifulSoup(response.text, 'html.parser')\n",
    "\n",
    "# Extract lesson links\n",
    "lesson_links = [a['href'] for a in soup.find_all('a', href=True)]\n",
    "\n",
    "# Base URL for plausible\n",
    "plausible_base_url = \"https://plausible.galaxyproject.eu/training.galaxyproject.org?period=all&page=\"\n",
    "\n",
    "# Filter only the lesson links that end with \"tutorial.html\"\n",
    "lesson_links = [link for link in lesson_links if link.endswith(\"tutorial.html\")]\n",
    "\n",
    "# Use a set to store unique plausible links\n",
    "unique_plausible_links = set()\n",
    "\n",
    "# Generate plausible links for the filtered lesson links and add to the set\n",
    "for link in lesson_links:\n",
    "    plausible_link = plausible_base_url + quote(link)\n",
    "    unique_plausible_links.add(plausible_link)\n",
    "\n",
    "# Print the unique plausible links\n",
    "for link in unique_plausible_links:\n",
    "    print(link)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
