{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Warning: SQLite3 version 3.40.0 and 3.41.2 have huge performance regressions; please install version 3.41.1 or 3.42!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from owlready2 import get_ontology, Thing\n",
    "import matplotlib.pyplot as plt\n",
    "import copy\n",
    "import numpy as np\n",
    "from adjustText import adjust_text\n",
    "from googleapiclient.discovery import build\n",
    "from datetime import timedelta\n",
    "import isodate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting Data from CoDex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "tutorials_url = \"../docs/supplementary/supplementary_table_5.tsv\"\n",
    "tools_url = \"../docs/supplementary/supplementary_table_3.tsv\"\n",
    "workflows_url = \"../docs/supplementary/supplementary_table_4.tsv\"\n",
    "\n",
    "tutorials_df = pd.read_csv(tutorials_url, sep=\"\\t\")\n",
    "tutorials_df = tutorials_df.drop_duplicates(keep=\"last\")\n",
    "\n",
    "tools_df = pd.read_csv(tools_url, sep=\"\\t\")\n",
    "\n",
    "workflows_df = pd.read_csv(workflows_url, sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up your API key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "API_KEY = ''  # Replace with your  youtube API key\n",
    "youtube = build('youtube', 'v3', developerKey=API_KEY)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorials Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new rows, one row for each EDAM term per tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### EDAM Operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Title</th>\n",
       "      <th>Link</th>\n",
       "      <th>EDAM topic</th>\n",
       "      <th>EDAM operation</th>\n",
       "      <th>Creation</th>\n",
       "      <th>Last modification</th>\n",
       "      <th>Version</th>\n",
       "      <th>Tutorial</th>\n",
       "      <th>Slides</th>\n",
       "      <th>...</th>\n",
       "      <th>Tools</th>\n",
       "      <th>Servers with precise tool versions</th>\n",
       "      <th>Servers with tool but different versions</th>\n",
       "      <th>Feedback number</th>\n",
       "      <th>Feedback mean note</th>\n",
       "      <th>Visitors</th>\n",
       "      <th>Page views</th>\n",
       "      <th>Visit duration</th>\n",
       "      <th>Video views</th>\n",
       "      <th>EDAM operation single</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Genomics, Microbiology</td>\n",
       "      <td>Sequence alignment, Sequence visualisation, Ma...</td>\n",
       "      <td>2018-06-14</td>\n",
       "      <td>2024-06-24</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fasta_filter_by_length, bedtools_sortbed, addV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6312</td>\n",
       "      <td>12503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sequence alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Genomics, Microbiology</td>\n",
       "      <td>Sequence alignment, Sequence visualisation, Ma...</td>\n",
       "      <td>2018-06-14</td>\n",
       "      <td>2024-06-24</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fasta_filter_by_length, bedtools_sortbed, addV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6312</td>\n",
       "      <td>12503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sequence visualisation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Genomics, Microbiology</td>\n",
       "      <td>Sequence alignment, Sequence visualisation, Ma...</td>\n",
       "      <td>2018-06-14</td>\n",
       "      <td>2024-06-24</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fasta_filter_by_length, bedtools_sortbed, addV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6312</td>\n",
       "      <td>12503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Mapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Genomics, Microbiology</td>\n",
       "      <td>Sequence alignment, Sequence visualisation, Ma...</td>\n",
       "      <td>2018-06-14</td>\n",
       "      <td>2024-06-24</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fasta_filter_by_length, bedtools_sortbed, addV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6312</td>\n",
       "      <td>12503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Genome visualisation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Genomics, Microbiology</td>\n",
       "      <td>Sequence alignment, Sequence visualisation, Ma...</td>\n",
       "      <td>2018-06-14</td>\n",
       "      <td>2024-06-24</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fasta_filter_by_length, bedtools_sortbed, addV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6312</td>\n",
       "      <td>12503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Read mapping</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Topic                                     Title  \\\n",
       "0  Assembly  Making sense of a newly assembled genome   \n",
       "1  Assembly  Making sense of a newly assembled genome   \n",
       "2  Assembly  Making sense of a newly assembled genome   \n",
       "3  Assembly  Making sense of a newly assembled genome   \n",
       "4  Assembly  Making sense of a newly assembled genome   \n",
       "\n",
       "                                                Link  \\\n",
       "0  https://training.galaxyproject.org//topics/ass...   \n",
       "1  https://training.galaxyproject.org//topics/ass...   \n",
       "2  https://training.galaxyproject.org//topics/ass...   \n",
       "3  https://training.galaxyproject.org//topics/ass...   \n",
       "4  https://training.galaxyproject.org//topics/ass...   \n",
       "\n",
       "                                  EDAM topic  \\\n",
       "0  Sequence assembly, Genomics, Microbiology   \n",
       "1  Sequence assembly, Genomics, Microbiology   \n",
       "2  Sequence assembly, Genomics, Microbiology   \n",
       "3  Sequence assembly, Genomics, Microbiology   \n",
       "4  Sequence assembly, Genomics, Microbiology   \n",
       "\n",
       "                                      EDAM operation    Creation  \\\n",
       "0  Sequence alignment, Sequence visualisation, Ma...  2018-06-14   \n",
       "1  Sequence alignment, Sequence visualisation, Ma...  2018-06-14   \n",
       "2  Sequence alignment, Sequence visualisation, Ma...  2018-06-14   \n",
       "3  Sequence alignment, Sequence visualisation, Ma...  2018-06-14   \n",
       "4  Sequence alignment, Sequence visualisation, Ma...  2018-06-14   \n",
       "\n",
       "  Last modification  Version Tutorial  Slides  ...  \\\n",
       "0        2024-06-24       29     True   False  ...   \n",
       "1        2024-06-24       29     True   False  ...   \n",
       "2        2024-06-24       29     True   False  ...   \n",
       "3        2024-06-24       29     True   False  ...   \n",
       "4        2024-06-24       29     True   False  ...   \n",
       "\n",
       "                                               Tools  \\\n",
       "0  fasta_filter_by_length, bedtools_sortbed, addV...   \n",
       "1  fasta_filter_by_length, bedtools_sortbed, addV...   \n",
       "2  fasta_filter_by_length, bedtools_sortbed, addV...   \n",
       "3  fasta_filter_by_length, bedtools_sortbed, addV...   \n",
       "4  fasta_filter_by_length, bedtools_sortbed, addV...   \n",
       "\n",
       "   Servers with precise tool versions  \\\n",
       "0                                 NaN   \n",
       "1                                 NaN   \n",
       "2                                 NaN   \n",
       "3                                 NaN   \n",
       "4                                 NaN   \n",
       "\n",
       "  Servers with tool but different versions Feedback number Feedback mean note  \\\n",
       "0                                      NaN               0                NaN   \n",
       "1                                      NaN               0                NaN   \n",
       "2                                      NaN               0                NaN   \n",
       "3                                      NaN               0                NaN   \n",
       "4                                      NaN               0                NaN   \n",
       "\n",
       "   Visitors  Page views  Visit duration  Video views   EDAM operation single  \n",
       "0      6312       12503             3.0            0      Sequence alignment  \n",
       "1      6312       12503             3.0            0  Sequence visualisation  \n",
       "2      6312       12503             3.0            0                 Mapping  \n",
       "3      6312       12503             3.0            0    Genome visualisation  \n",
       "4      6312       12503             3.0            0            Read mapping  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_tutorials_operation = []\n",
    "for idx, row in tutorials_df.iterrows():\n",
    "    terms = row[\"EDAM operation\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM operation single\"] = val\n",
    "            inflated_tutorials_operation.append(row_c)\n",
    "    else:\n",
    "        row_c = copy.copy(row_d)\n",
    "        row_c[\"EDAM operation single\"] = \"Undefined\"\n",
    "        inflated_tutorials_operation.append(row_c)\n",
    "\n",
    "df_inflated_tutorials_operation = pd.DataFrame(inflated_tutorials_operation)\n",
    "\n",
    "df_inflated_tutorials_operation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### EDAM topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Title</th>\n",
       "      <th>Link</th>\n",
       "      <th>EDAM topic</th>\n",
       "      <th>EDAM operation</th>\n",
       "      <th>Creation</th>\n",
       "      <th>Last modification</th>\n",
       "      <th>Version</th>\n",
       "      <th>Tutorial</th>\n",
       "      <th>Slides</th>\n",
       "      <th>...</th>\n",
       "      <th>Tools</th>\n",
       "      <th>Servers with precise tool versions</th>\n",
       "      <th>Servers with tool but different versions</th>\n",
       "      <th>Feedback number</th>\n",
       "      <th>Feedback mean note</th>\n",
       "      <th>Visitors</th>\n",
       "      <th>Page views</th>\n",
       "      <th>Visit duration</th>\n",
       "      <th>Video views</th>\n",
       "      <th>EDAM topic single</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Genomics, Microbiology</td>\n",
       "      <td>Sequence alignment, Sequence visualisation, Ma...</td>\n",
       "      <td>2018-06-14</td>\n",
       "      <td>2024-06-24</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fasta_filter_by_length, bedtools_sortbed, addV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6312</td>\n",
       "      <td>12503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sequence assembly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Genomics, Microbiology</td>\n",
       "      <td>Sequence alignment, Sequence visualisation, Ma...</td>\n",
       "      <td>2018-06-14</td>\n",
       "      <td>2024-06-24</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fasta_filter_by_length, bedtools_sortbed, addV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6312</td>\n",
       "      <td>12503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Making sense of a newly assembled genome</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Genomics, Microbiology</td>\n",
       "      <td>Sequence alignment, Sequence visualisation, Ma...</td>\n",
       "      <td>2018-06-14</td>\n",
       "      <td>2024-06-24</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fasta_filter_by_length, bedtools_sortbed, addV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6312</td>\n",
       "      <td>12503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Microbiology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Assembly of metagenomic sequencing data</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Metagenomics, Sequence assembly</td>\n",
       "      <td>Sequence trimming, Sequence assembly validatio...</td>\n",
       "      <td>2025-02-19</td>\n",
       "      <td>2025-02-19</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>ngsutils_bam_filter, cutadapt, bandage_image, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UseGalaxy.cz, UseGalaxy.eu</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Metagenomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Assembly of metagenomic sequencing data</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Metagenomics, Sequence assembly</td>\n",
       "      <td>Sequence trimming, Sequence assembly validatio...</td>\n",
       "      <td>2025-02-19</td>\n",
       "      <td>2025-02-19</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>ngsutils_bam_filter, cutadapt, bandage_image, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UseGalaxy.cz, UseGalaxy.eu</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sequence assembly</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Topic                                     Title  \\\n",
       "0  Assembly  Making sense of a newly assembled genome   \n",
       "1  Assembly  Making sense of a newly assembled genome   \n",
       "2  Assembly  Making sense of a newly assembled genome   \n",
       "3  Assembly   Assembly of metagenomic sequencing data   \n",
       "4  Assembly   Assembly of metagenomic sequencing data   \n",
       "\n",
       "                                                Link  \\\n",
       "0  https://training.galaxyproject.org//topics/ass...   \n",
       "1  https://training.galaxyproject.org//topics/ass...   \n",
       "2  https://training.galaxyproject.org//topics/ass...   \n",
       "3  https://training.galaxyproject.org//topics/ass...   \n",
       "4  https://training.galaxyproject.org//topics/ass...   \n",
       "\n",
       "                                  EDAM topic  \\\n",
       "0  Sequence assembly, Genomics, Microbiology   \n",
       "1  Sequence assembly, Genomics, Microbiology   \n",
       "2  Sequence assembly, Genomics, Microbiology   \n",
       "3            Metagenomics, Sequence assembly   \n",
       "4            Metagenomics, Sequence assembly   \n",
       "\n",
       "                                      EDAM operation    Creation  \\\n",
       "0  Sequence alignment, Sequence visualisation, Ma...  2018-06-14   \n",
       "1  Sequence alignment, Sequence visualisation, Ma...  2018-06-14   \n",
       "2  Sequence alignment, Sequence visualisation, Ma...  2018-06-14   \n",
       "3  Sequence trimming, Sequence assembly validatio...  2025-02-19   \n",
       "4  Sequence trimming, Sequence assembly validatio...  2025-02-19   \n",
       "\n",
       "  Last modification  Version Tutorial  Slides  ...  \\\n",
       "0        2024-06-24       29     True   False  ...   \n",
       "1        2024-06-24       29     True   False  ...   \n",
       "2        2024-06-24       29     True   False  ...   \n",
       "3        2025-02-19        0     True   False  ...   \n",
       "4        2025-02-19        0     True   False  ...   \n",
       "\n",
       "                                               Tools  \\\n",
       "0  fasta_filter_by_length, bedtools_sortbed, addV...   \n",
       "1  fasta_filter_by_length, bedtools_sortbed, addV...   \n",
       "2  fasta_filter_by_length, bedtools_sortbed, addV...   \n",
       "3  ngsutils_bam_filter, cutadapt, bandage_image, ...   \n",
       "4  ngsutils_bam_filter, cutadapt, bandage_image, ...   \n",
       "\n",
       "   Servers with precise tool versions  \\\n",
       "0                                 NaN   \n",
       "1                                 NaN   \n",
       "2                                 NaN   \n",
       "3                                 NaN   \n",
       "4                                 NaN   \n",
       "\n",
       "  Servers with tool but different versions Feedback number Feedback mean note  \\\n",
       "0                                      NaN               0                NaN   \n",
       "1                                      NaN               0                NaN   \n",
       "2                                      NaN               0                NaN   \n",
       "3               UseGalaxy.cz, UseGalaxy.eu               0                NaN   \n",
       "4               UseGalaxy.cz, UseGalaxy.eu               0                NaN   \n",
       "\n",
       "   Visitors  Page views  Visit duration  Video views  EDAM topic single  \n",
       "0      6312       12503             3.0            0  Sequence assembly  \n",
       "1      6312       12503             3.0            0           Genomics  \n",
       "2      6312       12503             3.0            0       Microbiology  \n",
       "3         1           1             0.0            0       Metagenomics  \n",
       "4         1           1             0.0            0  Sequence assembly  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_tutorials_topics = []\n",
    "for idx, row in tutorials_df.iterrows():\n",
    "    terms = row[\"EDAM topic\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM topic single\"] = val\n",
    "            inflated_tutorials_topics.append(row_c)\n",
    "    else:\n",
    "        row_c = copy.copy(row_d)\n",
    "        row_c[\"EDAM topic single\"] = \"Undefined\"\n",
    "        inflated_tutorials_topics.append(row_c)\n",
    "\n",
    "df_inflated_tutorials_topics = pd.DataFrame(inflated_tutorials_topics)\n",
    "\n",
    "df_inflated_tutorials_topics.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of unique tutorials: 36\n",
      "Total number of unique EDAM operations: 84\n",
      "Total number of unique EDAM topics: 25\n",
      "Total number of unique tutorials: 36\n",
      "Total number of tutorials with associated videos: 18\n",
      "Percentage of tutorials with videos: 50.00%\n",
      "Total number of videos: 18\n",
      "Total number of tutorials with associated workflows: 33\n",
      "Percentage of tutorials with workflows: 91.67%\n",
      "Total number of workflows: 33\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7418/3973245857.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2[\"Video Count\"] = df2[\"Video\"].apply(lambda x: len(x.split(',')) if isinstance(x, str) and x.strip() else 0)\n",
      "/tmp/ipykernel_7418/3973245857.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2[\"Workflow Count\"] = df2[\"Workflows\"].apply(lambda x: len(x.split(',')) if isinstance(x, str) and x.strip() else 0)\n"
     ]
    }
   ],
   "source": [
    "unique_tutorials = tutorials_df[\"Title\"].nunique()\n",
    "df2 = tutorials_df.drop_duplicates(subset='Title', keep=\"last\")\n",
    "unique_operations = df_inflated_tutorials_operation[\"EDAM operation single\"].nunique()\n",
    "unique_topics = df_inflated_tutorials_topics[\"EDAM topic single\"].nunique()\n",
    "\n",
    "# Count videos and workflows per tutorial with additional checks for string type\n",
    "df2[\"Video Count\"] = df2[\"Video\"].apply(lambda x: len(x.split(',')) if isinstance(x, str) and x.strip() else 0)\n",
    "df2[\"Workflow Count\"] = df2[\"Workflows\"].apply(lambda x: len(x.split(',')) if isinstance(x, str) and x.strip() else 0)\n",
    "\n",
    "# Calculate tutorials with associated videos and workflows (count of True values in each column)\n",
    "tutorials_with_videos = df2[\"Video\"].sum()\n",
    "tutorials_with_workflows = df2[\"Workflows\"].sum()\n",
    "\n",
    "# Calculate the total number of videos and workflows\n",
    "total_videos = df2[df2[\"Video\"] == True].shape[0]  # Total number of True values in Video column\n",
    "total_workflows = df2[df2[\"Workflows\"] == True].shape[0]  # Total number of True values in Workflows column\n",
    "\n",
    "# Calculate percentages\n",
    "percentage_with_videos = (tutorials_with_videos / unique_tutorials) * 100\n",
    "percentage_with_workflows = (tutorials_with_workflows / unique_tutorials) * 100\n",
    "\n",
    "# Print statistics\n",
    "print(f\"Total number of unique tutorials: {unique_tutorials}\")\n",
    "print(f\"Total number of unique EDAM operations: {unique_operations}\")\n",
    "print(f\"Total number of unique EDAM topics: {unique_topics}\")\n",
    "print(f\"Total number of unique tutorials: {unique_tutorials}\")\n",
    "print(f\"Total number of tutorials with associated videos: {tutorials_with_videos}\")\n",
    "print(f\"Percentage of tutorials with videos: {percentage_with_videos:.2f}%\")\n",
    "print(f\"Total number of videos: {total_videos}\")\n",
    "\n",
    "print(f\"Total number of tutorials with associated workflows: {tutorials_with_workflows}\")\n",
    "print(f\"Percentage of tutorials with workflows: {percentage_with_workflows:.2f}%\")\n",
    "print(f\"Total number of workflows: {total_workflows}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tutorials Videos on Youtube"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching video durations...\n",
      "\n",
      "\n",
      "Summary of Video Durations:\n",
      "Total number of unique videos: 17\n",
      "Total duration (hh:mm:ss): 15:42:09\n",
      "Updated table saved to tutorialsYoutube/table_with_api_video_durations.csv\n"
     ]
    }
   ],
   "source": [
    "# Function to get video duration using YouTube API\n",
    "def get_video_duration_api(video_url):\n",
    "    try:\n",
    "        # Extract the video ID from the URL\n",
    "        video_id = video_url.split('v=')[-1].split('&')[0] if 'youtube' in video_url else video_url.split('/')[-1]\n",
    "        request = youtube.videos().list(part=\"contentDetails\", id=video_id)\n",
    "        response = request.execute()\n",
    "\n",
    "        # Extract duration in ISO 8601 format\n",
    "        items = response.get(\"items\")\n",
    "        if items:\n",
    "            duration = items[0][\"contentDetails\"][\"duration\"]\n",
    "            # Convert ISO 8601 to total seconds\n",
    "            return parse_iso_duration(duration)\n",
    "        else:\n",
    "            print(f\"No details found for video: {video_url}\")\n",
    "            return 0\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing video {video_url}: {e}\")\n",
    "        return 0\n",
    "\n",
    "# Function to parse ISO 8601 duration to seconds\n",
    "\n",
    "def parse_iso_duration(duration):\n",
    "    duration_obj = isodate.parse_duration(duration)\n",
    "    return int(duration_obj.total_seconds())\n",
    "\n",
    "# Load the table containing YouTube links\n",
    "file_path = 'tutorialsYoutube/tutorialsVideos.csv'\n",
    "df = pd.read_csv(file_path)\n",
    "\n",
    "# Column containing YouTube links\n",
    "column_name = \"Video_Link\"\n",
    "\n",
    "# Apply function to get video durations\n",
    "print(\"Fetching video durations...\\n\")\n",
    "df['Duration_seconds'] = df[column_name].apply(get_video_duration_api)\n",
    "\n",
    "# Calculate total duration\n",
    "total_duration_seconds = int(df['Duration_seconds'].sum())  # Convert to Python int\n",
    "total_duration = timedelta(seconds=total_duration_seconds)\n",
    "\n",
    "# Display the total duration\n",
    "print(\"\\nSummary of Video Durations:\")\n",
    "print(f\"Total number of unique videos: {len(df)}\")\n",
    "print(f\"Total duration (hh:mm:ss): {total_duration}\")\n",
    "\n",
    "# Save updated table\n",
    "output_file = \"tutorialsYoutube/table_with_api_video_durations.csv\"\n",
    "df.to_csv(output_file, index=False)\n",
    "print(f\"Updated table saved to {output_file}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tools Statistics\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new rows, one row for each EDAM term per tool"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDAM Operation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "inflated_tools = []\n",
    "\n",
    "for idx, row in tools_df.iterrows():\n",
    "    terms = row[\"EDAM reduced operations\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM operation single\"] = val\n",
    "            inflated_tools.append(row_c)\n",
    "    else:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM operation single\"] = \"Undefined\"\n",
    "            inflated_tools.append(row_c)\n",
    "\n",
    "df_inflated_tools_operation = pd.DataFrame(inflated_tools)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDAM Topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "inflated_tools = []\n",
    "\n",
    "for idx, row in tools_df.iterrows():\n",
    "    terms = row[\"EDAM reduced topics\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM topic single\"] = val\n",
    "            inflated_tools.append(row_c)\n",
    "    else:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM topic single\"] = \"Undefined\"\n",
    "            inflated_tools.append(row_c)\n",
    "\n",
    "df_inflated_tools_topics = pd.DataFrame(inflated_tools)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(295,\n",
       " 21478831,\n",
       " 921291,\n",
       " 3123.020338983051,\n",
       " Suite ID                                                                             fastp\n",
       " Tool IDs                                                                             fastp\n",
       " Description                                  Fast all-in-one preprocessing for FASTQ files\n",
       " Suite first commit date                                                         2018-03-07\n",
       " Homepage                                                 https://github.com/OpenGene/fastp\n",
       "                                                                  ...                      \n",
       " Suite users on main servers                                                          69616\n",
       " Suite runs (last 5 years) on main servers                                          1773664\n",
       " Suite runs on main servers                                                         5325613\n",
       " To keep                                                                               True\n",
       " Deprecated                                                                           False\n",
       " Name: 76, Length: 75, dtype: object,\n",
       " Suite ID                                                                                mothur\n",
       " Tool IDs                                     mothur_align_check, mothur_align_seqs, mothur_...\n",
       " Description                                                                    Mothur wrappers\n",
       " Suite first commit date                                                             2016-06-24\n",
       " Homepage                                                                https://www.mothur.org\n",
       "                                                                    ...                        \n",
       " Suite users on main servers                                                             453202\n",
       " Suite runs (last 5 years) on main servers                                               995503\n",
       " Suite runs on main servers                                                             3253677\n",
       " To keep                                                                                   True\n",
       " Deprecated                                                                               False\n",
       " Name: 181, Length: 75, dtype: object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the total number of tools\n",
    "total_tools = tools_df.shape[0]\n",
    "\n",
    "# Calculate total tool usage over the last 5 years, over all main servers\n",
    "total_tool_usage = tools_df['Suite runs (last 5 years) on main servers'].sum()\n",
    "\n",
    "# Calculate total number of tool users over the last 5 years, over all main servers\n",
    "total_tool_users = tools_df['Suite users (last 5 years) on main servers'].sum()\n",
    "\n",
    "# Calculate the average number of tool users over the last 5 years, over all main servers\n",
    "average_tool_users = tools_df['Suite users (last 5 years) on main servers'].mean()\n",
    "\n",
    "# Find the most used tool over the last 5 years, over all main servers\n",
    "most_used_tool = tools_df.loc[tools_df['Suite runs (last 5 years) on main servers'].idxmax()]\n",
    "\n",
    "# Find the tool with the most used over the last 5 years, over all main servers\n",
    "most_users_tool = tools_df.loc[tools_df['Suite users (last 5 years) on main servers'].idxmax()]\n",
    "\n",
    "total_tools, total_tool_usage, total_tool_users, average_tool_users, most_used_tool, most_users_tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of tools: 294\n",
      "Total number of unique EDAM operations: 154\n",
      "Total number of unique EDAM topics: 95\n"
     ]
    }
   ],
   "source": [
    "unique_tools = tools_df[\"Suite ID\"].nunique()\n",
    "unique_tools_operations = df_inflated_tools_operation[\"EDAM operation single\"].nunique()\n",
    "unique_tools_topics = df_inflated_tools_topics[\"EDAM topic single\"].nunique()\n",
    "\n",
    "# Print statistics\n",
    "print(f\"Total number of tools: {unique_tools}\")\n",
    "print(f\"Total number of unique EDAM operations: {unique_tools_operations}\")\n",
    "print(f\"Total number of unique EDAM topics: {unique_tools_topics}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workflows Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new rows, one row for each EDAM term per workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDAM Operation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Source</th>\n",
       "      <th>ID</th>\n",
       "      <th>Link</th>\n",
       "      <th>Creators</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Creation time</th>\n",
       "      <th>Update time</th>\n",
       "      <th>Latest version</th>\n",
       "      <th>Versions</th>\n",
       "      <th>Number of steps</th>\n",
       "      <th>Tools</th>\n",
       "      <th>EDAM operations</th>\n",
       "      <th>EDAM topics</th>\n",
       "      <th>License</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Projects</th>\n",
       "      <th>To keep</th>\n",
       "      <th>Deprecated</th>\n",
       "      <th>EDAM operation single</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-complete/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1274</td>\n",
       "      <td>https://workflowhub.eu/workflows/1274?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>, CONVERTER_uncompressed_to_gz, __MERGE_COLLEC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-its/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1273</td>\n",
       "      <td>https://workflowhub.eu/workflows/1273?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>, __FILTER_FROM_FILE__, biom_convert, taxonomy...</td>\n",
       "      <td>Mapping, k-mer counting, Visualisation, Format...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Mapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-its/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1273</td>\n",
       "      <td>https://workflowhub.eu/workflows/1273?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>, __FILTER_FROM_FILE__, biom_convert, taxonomy...</td>\n",
       "      <td>Mapping, k-mer counting, Visualisation, Format...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>k-mer counting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-its/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1273</td>\n",
       "      <td>https://workflowhub.eu/workflows/1273?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>, __FILTER_FROM_FILE__, biom_convert, taxonomy...</td>\n",
       "      <td>Mapping, k-mer counting, Visualisation, Format...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Visualisation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-its/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1273</td>\n",
       "      <td>https://workflowhub.eu/workflows/1273?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>, __FILTER_FROM_FILE__, biom_convert, taxonomy...</td>\n",
       "      <td>Mapping, k-mer counting, Visualisation, Format...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Formatting</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Name       Source    ID  \\\n",
       "0  mgnify-amplicon-pipeline-v5-complete/main  WorkflowHub  1274   \n",
       "1       mgnify-amplicon-pipeline-v5-its/main  WorkflowHub  1273   \n",
       "2       mgnify-amplicon-pipeline-v5-its/main  WorkflowHub  1273   \n",
       "3       mgnify-amplicon-pipeline-v5-its/main  WorkflowHub  1273   \n",
       "4       mgnify-amplicon-pipeline-v5-its/main  WorkflowHub  1273   \n",
       "\n",
       "                                              Link                 Creators  \\\n",
       "0  https://workflowhub.eu/workflows/1274?version=1  Rand Zoabi, Paul Zierep   \n",
       "1  https://workflowhub.eu/workflows/1273?version=1  Rand Zoabi, Paul Zierep   \n",
       "2  https://workflowhub.eu/workflows/1273?version=1  Rand Zoabi, Paul Zierep   \n",
       "3  https://workflowhub.eu/workflows/1273?version=1  Rand Zoabi, Paul Zierep   \n",
       "4  https://workflowhub.eu/workflows/1273?version=1  Rand Zoabi, Paul Zierep   \n",
       "\n",
       "  Tags Creation time Update time  Latest version  Versions  Number of steps  \\\n",
       "0  NaN    2025-02-04  2025-02-04               1         1             20.0   \n",
       "1  NaN    2025-02-04  2025-02-04               1         1             30.0   \n",
       "2  NaN    2025-02-04  2025-02-04               1         1             30.0   \n",
       "3  NaN    2025-02-04  2025-02-04               1         1             30.0   \n",
       "4  NaN    2025-02-04  2025-02-04               1         1             30.0   \n",
       "\n",
       "                                               Tools  \\\n",
       "0  , CONVERTER_uncompressed_to_gz, __MERGE_COLLEC...   \n",
       "1  , __FILTER_FROM_FILE__, biom_convert, taxonomy...   \n",
       "2  , __FILTER_FROM_FILE__, biom_convert, taxonomy...   \n",
       "3  , __FILTER_FROM_FILE__, biom_convert, taxonomy...   \n",
       "4  , __FILTER_FROM_FILE__, biom_convert, taxonomy...   \n",
       "\n",
       "                                     EDAM operations EDAM topics     License  \\\n",
       "0                                                NaN         NaN  Apache-2.0   \n",
       "1  Mapping, k-mer counting, Visualisation, Format...         NaN  Apache-2.0   \n",
       "2  Mapping, k-mer counting, Visualisation, Format...         NaN  Apache-2.0   \n",
       "3  Mapping, k-mer counting, Visualisation, Format...         NaN  Apache-2.0   \n",
       "4  Mapping, k-mer counting, Visualisation, Format...         NaN  Apache-2.0   \n",
       "\n",
       "   DOI                                 Projects  To keep  Deprecated  \\\n",
       "0  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "1  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "2  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "3  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "4  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "\n",
       "  EDAM operation single  \n",
       "0             Undefined  \n",
       "1               Mapping  \n",
       "2        k-mer counting  \n",
       "3         Visualisation  \n",
       "4            Formatting  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_workflows_operation = []\n",
    "for idx, row in workflows_df.iterrows():\n",
    "    terms = row[\"EDAM operations\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM operation single\"] = val\n",
    "            inflated_workflows_operation.append(row_c)\n",
    "    else:\n",
    "        row_c = copy.copy(row_d)\n",
    "        row_c[\"EDAM operation single\"] = \"Undefined\"\n",
    "        inflated_workflows_operation.append(row_c)\n",
    "\n",
    "df_inflated_workflows_operation = pd.DataFrame(inflated_workflows_operation)\n",
    "\n",
    "df_inflated_workflows_operation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDAM topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Source</th>\n",
       "      <th>ID</th>\n",
       "      <th>Link</th>\n",
       "      <th>Creators</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Creation time</th>\n",
       "      <th>Update time</th>\n",
       "      <th>Latest version</th>\n",
       "      <th>Versions</th>\n",
       "      <th>Number of steps</th>\n",
       "      <th>Tools</th>\n",
       "      <th>EDAM operations</th>\n",
       "      <th>EDAM topics</th>\n",
       "      <th>License</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Projects</th>\n",
       "      <th>To keep</th>\n",
       "      <th>Deprecated</th>\n",
       "      <th>EDAM topic single</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-complete/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1274</td>\n",
       "      <td>https://workflowhub.eu/workflows/1274?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>, CONVERTER_uncompressed_to_gz, __MERGE_COLLEC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-its/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1273</td>\n",
       "      <td>https://workflowhub.eu/workflows/1273?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>, __FILTER_FROM_FILE__, biom_convert, taxonomy...</td>\n",
       "      <td>Mapping, k-mer counting, Visualisation, Format...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-quality-control-pa...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1272</td>\n",
       "      <td>https://workflowhub.eu/workflows/1272?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>17.0</td>\n",
       "      <td>trimmomatic, fastq_filter, __UNZIP_COLLECTION_...</td>\n",
       "      <td>Sequence composition calculation, Statistical ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-quality-control-si...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1271</td>\n",
       "      <td>https://workflowhub.eu/workflows/1271?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>14.0</td>\n",
       "      <td>trimmomatic, fastq_filter, cshl_fasta_formatte...</td>\n",
       "      <td>Sequencing quality control, Read pre-processin...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mgnify-amplicon-pipeline-v5-rrna-prediction/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1270</td>\n",
       "      <td>https://workflowhub.eu/workflows/1270?version=1</td>\n",
       "      <td>Rand Zoabi, Paul Zierep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>2025-02-04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>47.0</td>\n",
       "      <td>, __FILTER_FROM_FILE__, infernal_cmsearch, bio...</td>\n",
       "      <td>Comparison, Mapping, Nucleic acid feature dete...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Name       Source    ID  \\\n",
       "0          mgnify-amplicon-pipeline-v5-complete/main  WorkflowHub  1274   \n",
       "1               mgnify-amplicon-pipeline-v5-its/main  WorkflowHub  1273   \n",
       "2  mgnify-amplicon-pipeline-v5-quality-control-pa...  WorkflowHub  1272   \n",
       "3  mgnify-amplicon-pipeline-v5-quality-control-si...  WorkflowHub  1271   \n",
       "4   mgnify-amplicon-pipeline-v5-rrna-prediction/main  WorkflowHub  1270   \n",
       "\n",
       "                                              Link                 Creators  \\\n",
       "0  https://workflowhub.eu/workflows/1274?version=1  Rand Zoabi, Paul Zierep   \n",
       "1  https://workflowhub.eu/workflows/1273?version=1  Rand Zoabi, Paul Zierep   \n",
       "2  https://workflowhub.eu/workflows/1272?version=1  Rand Zoabi, Paul Zierep   \n",
       "3  https://workflowhub.eu/workflows/1271?version=1  Rand Zoabi, Paul Zierep   \n",
       "4  https://workflowhub.eu/workflows/1270?version=1  Rand Zoabi, Paul Zierep   \n",
       "\n",
       "  Tags Creation time Update time  Latest version  Versions  Number of steps  \\\n",
       "0  NaN    2025-02-04  2025-02-04               1         1             20.0   \n",
       "1  NaN    2025-02-04  2025-02-04               1         1             30.0   \n",
       "2  NaN    2025-02-04  2025-02-04               1         1             17.0   \n",
       "3  NaN    2025-02-04  2025-02-04               1         1             14.0   \n",
       "4  NaN    2025-02-04  2025-02-04               1         1             47.0   \n",
       "\n",
       "                                               Tools  \\\n",
       "0  , CONVERTER_uncompressed_to_gz, __MERGE_COLLEC...   \n",
       "1  , __FILTER_FROM_FILE__, biom_convert, taxonomy...   \n",
       "2  trimmomatic, fastq_filter, __UNZIP_COLLECTION_...   \n",
       "3  trimmomatic, fastq_filter, cshl_fasta_formatte...   \n",
       "4  , __FILTER_FROM_FILE__, infernal_cmsearch, bio...   \n",
       "\n",
       "                                     EDAM operations EDAM topics     License  \\\n",
       "0                                                NaN         NaN  Apache-2.0   \n",
       "1  Mapping, k-mer counting, Visualisation, Format...         NaN  Apache-2.0   \n",
       "2  Sequence composition calculation, Statistical ...         NaN  Apache-2.0   \n",
       "3  Sequencing quality control, Read pre-processin...         NaN  Apache-2.0   \n",
       "4  Comparison, Mapping, Nucleic acid feature dete...         NaN  Apache-2.0   \n",
       "\n",
       "   DOI                                 Projects  To keep  Deprecated  \\\n",
       "0  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "1  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "2  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "3  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "4  NaN  Intergalactic Workflow Commission (IWC)     True       False   \n",
       "\n",
       "  EDAM topic single  \n",
       "0         Undefined  \n",
       "1         Undefined  \n",
       "2         Undefined  \n",
       "3         Undefined  \n",
       "4         Undefined  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_workflows_topics = []\n",
    "for idx, row in workflows_df.iterrows():\n",
    "    terms = row[\"EDAM topics\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM topic single\"] = val\n",
    "            inflated_workflows_topics.append(row_c)\n",
    "    else:\n",
    "        row_c = copy.copy(row_d)\n",
    "        row_c[\"EDAM topic single\"] = \"Undefined\"\n",
    "        inflated_workflows_topics.append(row_c)\n",
    "\n",
    "df_inflated_workflows_topics = pd.DataFrame(inflated_workflows_topics)\n",
    "\n",
    "df_inflated_workflows_topics.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of workflows: 96\n",
      "Total number of unique EDAM operations: 110\n",
      "Total number of unique EDAM topics: 9\n"
     ]
    }
   ],
   "source": [
    "unique_workflows = workflows_df[\"Name\"].nunique()\n",
    "unique_workflows_operations = df_inflated_workflows_operation[\"EDAM operation single\"].nunique()\n",
    "unique_workflows_topics = df_inflated_workflows_topics[\"EDAM topic single\"].nunique()\n",
    "\n",
    "# Print statistics\n",
    "print(f\"Total number of workflows: {unique_workflows}\")\n",
    "print(f\"Total number of unique EDAM operations: {unique_workflows_operations}\")\n",
    "print(f\"Total number of unique EDAM topics: {unique_workflows_topics}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
